rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,223 @@
+
+Part of the mechanics requires [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+or a similar mechanism.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Why should we do this?
+
+It is desirable to know which inputs yielded an output, as this knowledge is central to tracing the preceding artifact
+inputs which ultimately result in said output.
+
+While the necessary information to do so may exist on the stamped resources, and thus could be inferred from inspecting
+the resource, Cartographer does not provide an agnostic means to leverage this. This means users must delve into these
+resources and make these inferences themselves. Doing so requires more explicit knowledge of exploring resources, is
+laborious, and error-prone.
+
+While this RFC does not directly propose UX changes to enable this, it describes mechanisms that enable Cartographer to
+internally recognize which inputs yield a given output, which can be used as the basis for UX changes.
+
+## What use cases does it support?
+
+The changes proposed here do not surface the correlation between inputs and outputs, but they introduce mechanics which
+will enable us to build features which do so.
+
+An interesting side effect of these mechanics is that Cartographer will not propagate artifacts output by a resource to
+the next resource unless they can be correlated to input provided to the resource by Cartographer. This is desirable
+from a security perspective, and the outcomes of this behavior are explored in detail in the
+[Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+# What it is
+
+[what-it-is]: #what-it-is
+
+New fields are added to templates which allow template authors to specify correlation rules using `correlationRules`,
+which informs Cartographer of how to compare the resource's output (published in the owner's status) with its inputs
+(based on where the input values were substituted into the resource stamped bt template). Where the output does not
+expose useful matching information, it is possible to match against the resource's `spec` instead, with some caveats.
+
+Template authors are thought to be the ideal target, as they are likely to have the most intimate knowledge of the
+resource that their templates stamp out.
+
+Example matching inputs against outputs with the `correlationRules` field:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+  name: example-source-scanning-thing
+spec:
+  template:
+    apiVersion: my-scanner/v1
+    kind: SourceScan
+    metadata:
+      name: ...
+    spec:
+      scanUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+correlationRules:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(status.outputs.url)"             #evaluated against resource status
```

Comment:
a few minor changes in this one suggestion:
- `input` and `output` are overloaded for resource artifacts. Using `expected` and `actual`, other terms can fit as well
- the "input" is setting a expectation for the value to find, updated to use the regular cartographer template syntax
- the output is a path on the stamped object, dropping the template syntax since we don't use it on other values read from the stamped object.

```suggestion
  - expectedValue: "$(sources.source.url)$"              #evaluated against inputs in template context
    actualPath: ".status.outputs.url"             #evaluated against the in-cluster resource
```